lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+  }
+
+  /// \brief Reference to a labeled FORMAT statement
+  void AddFmtBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().formatStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief Reference to a labeled statement as a (possible) branch
+  void AddBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().otherStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief References to labeled statements as (possible) branches
+  void AddBase(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      AddBase(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;  ///< results for each program unit
+  ErrorHandler eh;  ///< error handler, collects messages
+  parser::CharBlock currentPosition_{
+      nullptr};  ///< current location in parse tree
+  ScopeProxy currentScope_{0};  ///< current scope in the model
+  std::vector<std::string> constructNames_;
+};
+
+template<typename A, typename B>
+bool InInclusiveScope(const A &scopes, B tail, const B &head) {
+  assert(HasScope(head));
+  assert(HasScope(tail));
+  while (HasScope(tail) && (tail != head)) {
+    tail = scopes[tail];
+  }
+  return tail == head;
+}
+
+ParseTreeAnalyzer LabelAnalysis(const parser::Program &program) {
+  ParseTreeAnalyzer analysis;
+  Walk(program, analysis);
+  return analysis;
+}
+
+template<typename A, typename B>
+inline bool InBody(const A &position, const B &pair) {
+  assert(pair.first.begin() < pair.second.begin());
+  return (position.begin() >= pair.first.begin()) &&
+      (position.begin() < pair.second.end());
+}
+
+template<typename A, typename B>
+LabelStmtInfo GetLabel(const A &labels, const B &label) {
+  const auto iter{labels.find(label)};
+  if (iter == labels.cend()) {
+    return {0u, nullptr, TargetStmtType{}};
+  } else {
+    return iter->second;
+  }
+}
+
+/// \brief Check branches into a <i>label-do-stmt</i>
+/// Relates to 11.1.7.3, loop activation
+template<typename A, typename B, typename C, typename D>
+inline void CheckBranchesIntoDoBody(const A &branches, const B &labels,
+    const C &scopes, const D &loopBodies, ErrorHandler &eh) {
+  for (const auto branch : branches) {
+    const auto &label{std::get<parser::Label>(branch)};
+    auto branchTarget{GetLabel(labels, label)};
+    if (HasScope(std::get<ScopeProxy>(branchTarget))) {
+      const auto &fromPosition{std::get<parser::CharBlock>(branch)};
+      const auto &toPosition{std::get<parser::CharBlock>(branchTarget)};
+      for (const auto body : loopBodies) {
+        if (!InBody(fromPosition, body) && InBody(toPosition, body)) {
+          // this is an error: branch into labeled DO body
+          if (isStrictF18) {
+            Report(eh, fromPosition,
+                parser::MessageFormattedText{
+                    "branch into '%s' from another scope"_err_en_US,
```

Comment:
Does this have to be an error?  Maybe.  See comment above.

Reply:
reworked